src/main/kotlin/org/rust/lang/core/macros/RangeMap.kt

```diff
@@ -0,0 +1,177 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.macros
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.util.SmartList
+import org.rust.lang.core.psi.RsMacroArgument
+import org.rust.lang.core.psi.RsMacroCall
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.expansion
+import org.rust.lang.core.psi.ext.startOffset
+import org.rust.stdext.optimizeList
+import org.rust.stdext.readVarInt
+import org.rust.stdext.writeVarInt
+import java.io.DataInputStream
+import java.io.DataOutputStream
+
+// TODO optimize it using sorted array and binary search
+class RangeMap private constructor(private val ranges: List<Pair<TextRange, TextRange>>) {
```

Comment:
Let's add javadoc about `ranges` because it's not clear what these ranges mean without reading their usages